nntrainer/layers/rnn.cpp

```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void RNNLayer::calcDerivative() {
-  //  NYI
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &weight =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &ret_ = net_input[0]->getGradientRef();
+
+  derivative_.dot(weight, ret_, false, true);
 }
 
 void RNNLayer::calcGradient() {
-  // NYI
+  Tensor &djdw_x =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getGradientRef();
+  Tensor &djdw_h =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getGradientRef();
+  Tensor &djdb_h =
+    weightAt(static_cast<int>(RNNParams::bias_h)).getGradientRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getVariableRef();
+
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor dh_nx = Tensor(TensorDim(1, 1, 1, derivative_.width()));
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor deriv_t = derivative_.getBatchSlice(b, 1);
+    Tensor xs_t = input_.getBatchSlice(b, 1);
+    Tensor hs_t = hidden_.getBatchSlice(b, 1);
+    dh_nx.setZero();
+
+    Tensor dh;
+    Tensor xs;
+    Tensor hs_prev;
+    Tensor hs;
+
+    for (unsigned int t = deriv_t.height(); t-- > 0;) {
+      dh = deriv_t.getSharedDataTensor(TensorDim(1, 1, 1, deriv_t.width()),
+                                       t * deriv_t.width());
+      xs = xs_t.getSharedDataTensor(TensorDim(1, 1, 1, xs_t.width()),
+                                    t * xs_t.width());
+      hs = hs_t.getSharedDataTensor(TensorDim(1, 1, 1, hs_t.width()),
+                                    t * hs_t.width());
+      if (t == 0) {
+        hs_prev = Tensor(TensorDim(1, 1, 1, hs_t.width()));
+        hs_prev.setZero();
+      } else {
+        hs_prev = hs_t.getSharedDataTensor(TensorDim(1, 1, 1, hs_t.width()),
+                                           (t - 1) * hs_t.width());
+      }
+
+      if (t < deriv_t.height() - 1) {
+        dh.add_i(dh_nx);
+      }
+
+      acti_func.run_prime_fn(hs, dh, dh);
+      dh.multiply_i(hs);
+      djdb_h.add_i(dh);
+
+      djdw_x.add_i(xs.dot(dh, true, false));
```

Comment:
how about set replace add_i with providing the last parameter `beta` as 1.0 and initialize the djdw_x and djdw_h as zero?

Reply:
right. I forget to set the gradients zero. I will add. 
>> add_i with providing the last parameter beta as 1.0
so.. you mean A= beta*A + B and if beta is 0.0 the A is zero??

Reply:
> so.. you mean A= beta*A + B and if beta is 0.0 the A is zero??

right. If beta is 0.0, then the original value of A will not affect the result.

Reply:
It's good. I will update.